_posts/2020-06-23-Durr_Hoyer_Post_1.md

```diff
@@ -0,0 +1,328 @@
+            {
+                (ControlledOnInt(1,Z))(register!,aux); // If qubit is 1 flip it!
+            }
+        }
+    }
+
+    
+---
+
+For futher information on how this was derived take a look at 'Tight bounds on quantum searching' [2].
+It is important to not the above algorithm only works for a table with even entries. 
+The algorithm breaks down when applying the Hadamard gate as the Hadamard is layed across the diagnoal of a identity matrix which is equal in dimensions to the number of qubits we have. 
+With a bit of math, if we try to lay a 2x2 matrix along an odd dimensioned identity matrix the transformation is not retained.
+To circumvent this we introduce the following implementation, utilizing QFT.
+
+        operation Algorithm_Odd(TableLength : Int, RandomIndex : Int) : Unit
+        {
+            using ((Register) = (Qubit[TableLength])) // intialize register to number of qubits as there are indices
+            {
+                within
+                    {   
+                        let Marker = Register[RandomIndex]; // Grab the qubit in the RandomIndex and set it aside
+
+                        PrepareUniformSuperposition(TableLength,LittleEndian(Register)); // Create Uniform Superposition of all indices
+
+                        Controlled Z(Register,Marker); // Apply Oracle to flip all states that are T[j]<T[y]
+                    }
+                apply 
+                    {
+                        QFTLE(LittleEndian(register)); // Implemntation for odd number of table entries
+
+                        ApplyConditionalPhase_0(LittleEndian(Register)); // Reflect qubits that are 0s
+
+                        QFT(BigEndian(register)); // Inverse QFT by using BigEndian
+
+                        ApplyConditionalPhase(LittleEndian(Register)); //Reflect qubits that are 1s
+                    }
+            }
+        }
+
+Now we refer back to our QMSA outline to observe that the Algorithm(TableLenght,RandomIndex) is iterated on until we find a suitable y' or we simply hit our time limit. The true stars of this algorithm are the time limit, which guarantees O(sqrt(N)), the generalized Grovers algorithm given in QESA, which provides for easy implementation and has O(1) for each iteration, and lastly, the oracle function which marks our states, which along with our intialization of qubits has O(log(n)).
+
+Here is the python host that will apply the conditions of QMSA while using QESA.
+
+---
+
+    class DH(object):
+
+      def __init__(self,table):
+
+          self.table = np.ndarray.flatten(table) # Flatten table
+
+          self.N= len(self.table) # Number of elements in the table
+
+          self.y = int(np.random.uniform(0,self.N-1)) # Choose our random index
+
+      def QMSA(self,N,y,table):
+      
+          N = self.N
+
+          y = self.y
+
+          t = self.table
+
+          y_prime = N
+
+          time_limit = (22.5*np.sqrt(N)+1.4*np.log(N)**2) # Time limit as specified in QMSA

```

Comment:
What if it runs on a slow machine?

Reply:
This is a good point. Maybe a docker env should be implemented to provide for consistent results... or we can allow for users alleviate the time limit so they can run it on a slower computer and still get the 50% accuracy as talked of in the paper.